### PR TITLE
FM2-185:implement the immunization resource

### PR DESCRIPTION
**Pull request by suruchee**
I worked on the Implementation of Integration test in immunization resource

### CODE REVIEW

File: integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r4/ImmunizationFhirResourceProviderIntegrationTest.java [hunk 1]
```diff
@@ -0,0 +1,505 @@
...
+		
+		assertThat(Immunization, notNullValue());
+		assertThat(Immunization.getIdElement().getIdPart(), equalTo(IMMUNIZATION_UUID));
+		assertThat(Immunization, validResource());
+	}
+	
+	@Test
+	public void shouldReturnNotFoundWhenIMMUNIZATIONNotFoundAsJson() throws Exception {
+		MockHttpServletResponse response = get("/Immunization/" + UNKNOWN_IMMUNIZATION_UUID).accept(FhirMediaTypes.JSON)
+		        .go();
```
Comment (HerbertYiga) [inline integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r4/ImmunizationFhirResourceProviderIntegrationTest.java:88]: Add this to the same line above

**Comment by HerbertYiga** [comment]
Am also seeing an error line org.dbunit.dataset.NoSuchTableException: immunization_history in your maven build logs

File: test-data/src/test/resources/org/openmrs/module/fhir2/providers/ImmunizationWebTest_create.json [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+		"actor": {
+			"reference": "Practitioner/f9badd80-ab76-11e2-9e96-0800200c9a66 ",
+			"type": "Practitioner"
+		}
+	},
+	"protocolApplied": [{
+		"doseNumberPositiveInt": 2,
+		"series": "Dose "
+	}]
+}
```
Comment (ibacher) [inline test-data/src/test/resources/org/openmrs/module/fhir2/providers/ImmunizationWebTest_create.json:37]:
```suggestion
}

```

File: test-data/src/test/resources/org/openmrs/module/fhir2/providers/ImmunizationWebTest_create.xml [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+			<code value="984AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"/>
+		</coding>
+	</code>
+	<patient>
+		<reference value="Patient/da7f524f-27ce-4bb2-86d6-6d1d05312bd5"/>
+	</patient>
+	<recorder>
+		<reference value="Practitioner/c98a1558-e131-11de-babe-001e378eb67e"/>
+	</recorder>
+</Immunization>
```
Comment (ibacher) [inline test-data/src/test/resources/org/openmrs/module/fhir2/providers/ImmunizationWebTest_create.xml:23]:
```suggestion
</Immunization>

```

File: test-data/src/test/resources/org/openmrs/module/fhir2/api/dao/impl/FhirImmunizationDaoImplTest_initial_data.xml [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    <concept_name concept_id="70012" name="VACCINE MANUFACTURER" locale="en" creator="1" date_created="2020-07-07 19:30:26.0" concept_name_id="70012" concept_name_type="FULLY_SPECIFIED" locale_preferred="1" voided="false" uuid="81cf6b39-55c1-4636-ab5d-bf52a7f77b7e"/>
+    <concept_name concept_id="70013" name="VACCINE LOT NUMBER" locale="en" creator="1" date_created="2020-07-07 19:30:26.0" concept_name_id="70013" concept_name_type="FULLY_SPECIFIED" locale_preferred="1" voided="false" uuid="d611caa1-d5bd-402b-9f9c-bfdeee6e7ed9"/>
+    <concept_name concept_id="70014" name="Vaccine lot expiration date" locale="en" creator="1" date_created="2020-07-07 19:30:26.0" concept_name_id="70014" concept_name_type="FULLY_SPECIFIED" locale_preferred="1" voided="false" uuid="a349f90e-f317-4817-ae53-e9430e857f8d"/>
+
+    <immunization_history immunization_history_id="1421" patient_id="2" immunization_concept_id="1410" creator="1"  date_changed="2020-07-07 19:08:19.0" date_created="2020-07-07 19:30:26.0" voided="true" uuid="1421AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"/>
+
+    
+</dataset>
```
Comment (ibacher) [inline test-data/src/test/resources/org/openmrs/module/fhir2/api/dao/impl/FhirImmunizationDaoImplTest_initial_data.xml:43]:
```suggestion   
</dataset>

```

File: test-data/src/test/resources/org/openmrs/module/fhir2/api/dao/impl/FhirImmunizationDaoImplTest_initial_data.xml [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    <person_name person_name_id="2" preferred="true" person_id="6" given_name="John" middle_name="F" family_name="Doe" creator="1" date_created="2005-01-01 00:00:00.0" voided="false" uuid="7e2acadc-5073-4a39-914a-debcbec8c1c9"/>
+    <person_name person_name_id="3" preferred="true" person_id="7" given_name="I" middle_name="am" family_name="voided" creator="1" date_created="2005-01-01 00:00:00.0" voided="false" uuid="37e60810-0ba9-4f32-ba80-0af59f59e966"/>
+    <patient patient_id="6" creator="1" date_created="2005-01-01 00:00:00.0" voided="false"/>
+    <patient patient_id="7" creator="1" date_created="2005-01-01 00:00:00.0" voided="true"/>
+
+    <patient_identifier patient_identifier_id="1" patient_id="6" identifier="M4001-1" identifier_type="1" preferred="0" location_id="1" creator="1" date_created="2004-01-01 00:00:00.0" voided="false" uuid="504c83c7-cfbf-4ae7-a4da-bdfa3236689f"/>
+    <patient_identifier patient_identifier_id="2" patient_id="7" identifier="M4001-2" identifier_type="1" preferred="0" location_id="1" creator="1" date_created="2004-01-01 00:00:00.0" voided="false" uuid="504c83c7-cfbf-4ae7-a4da-bdfa3236689g"/>
+
+   
+
```
Comment (ibacher) [inline test-data/src/test/resources/org/openmrs/module/fhir2/api/dao/impl/FhirImmunizationDaoImplTest_initial_data.xml:31]:
```suggestion

```

File: integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r4/ImmunizationFhirResourceProviderIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,505 @@
...
+		
+		entries = results.getEntry();
+		
+		assertThat(entries, everyItem(hasResource(hasProperty("patient",
+		    hasProperty("referenceElement", hasProperty("idPart", equalTo("8d703ff2-c3e2-4070-9737-73e713d5a50d")))))));
+		assertThat(entries, everyItem(hasResource(validResource())));
+		
+	}
+	
+}
```
Comment (ibacher) [inline integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r4/ImmunizationFhirResourceProviderIntegrationTest.java:505]:
```suggestion
}

```

File: integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r4/ImmunizationFhirResourceProviderIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,505 @@
...
+		assertThat(results, notNullValue());
+		assertThat(results.getType(), equalTo(Bundle.BundleType.SEARCHSET));
+		assertThat(results.hasEntry(), is(true));
+		
+		entries = results.getEntry();
+		
+		assertThat(entries, everyItem(hasResource(hasProperty("patient",
+		    hasProperty("referenceElement", hasProperty("idPart", equalTo("8d703ff2-c3e2-4070-9737-73e713d5a50d")))))));
+		assertThat(entries, everyItem(hasResource(validResource())));
+
```
Comment (ibacher) [inline integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r4/ImmunizationFhirResourceProviderIntegrationTest.java:502]:
```suggestion
```

File: integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r4/ImmunizationFhirResourceProviderIntegrationTest.java [hunk 8]
```diff
@@ -0,0 +1,505 @@
...
```
Comment (ibacher) [inline integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r4/ImmunizationFhirResourceProviderIntegrationTest.java:461]:
```suggestion
```

File: integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r4/ImmunizationFhirResourceProviderIntegrationTest.java [hunk 9]
```diff
@@ -0,0 +1,505 @@
...
+		
+		Immunization Immunization = readResponse(response);
+		
+		assertThat(Immunization, notNullValue());
+		assertThat(Immunization.getIdElement().getIdPart(), equalTo(IMMUNIZATION_UUID));
+		assertThat(Immunization, validResource());
+	}
+	
+	@Test
+	public void shouldReturnNotFoundWhenIMMUNIZATIONNotFoundAsXML() throws Exception {
```
Comment (ibacher) [inline integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r4/ImmunizationFhirResourceProviderIntegrationTest.java:116]:
```suggestion
	public void shouldReturnNotFoundWhenImmunizationNNotFoundAsXML() throws Exception {
```

File: integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r4/ImmunizationFhirResourceProviderIntegrationTest.java [hunk 10]
```diff
@@ -0,0 +1,505 @@
...
+		assertThat(response.getContentAsString(), notNullValue());
+		
+		OperationOutcome operationOutcome = readOperationOutcome(response);
+		
+		assertThat(operationOutcome, notNullValue());
+		assertThat(operationOutcome.hasIssue(), is(true));
+	}
+	
+	@Test
+	public void shouldReturnExistingIMMUNIZATIONAsXML() throws Exception {
```
Comment (ibacher) [inline integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r4/ImmunizationFhirResourceProviderIntegrationTest.java:101]:
```suggestion
	public void shouldReturnExistingImmunizationAsXML() throws Exception {
```

File: integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r4/ImmunizationFhirResourceProviderIntegrationTest.java [hunk 11]
```diff
@@ -0,0 +1,505 @@
...
+		
+		Immunization Immunization = readResponse(response);
+		
+		assertThat(Immunization, notNullValue());
+		assertThat(Immunization.getIdElement().getIdPart(), equalTo(IMMUNIZATION_UUID));
+		assertThat(Immunization, validResource());
+	}
+	
+	@Test
+	public void shouldReturnNotFoundWhenIMMUNIZATIONNotFoundAsJson() throws Exception {
```
Comment (ibacher) [inline integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r4/ImmunizationFhirResourceProviderIntegrationTest.java:86]:
```suggestion
	public void shouldReturnNotFoundWhenImmunizationNotFoundAsJson() throws Exception {
```

File: integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r4/ImmunizationFhirResourceProviderIntegrationTest.java [hunk 12]
```diff
@@ -0,0 +1,505 @@
...
+	
+	@Before
+	@Override
+	public void setup() throws Exception {
+		super.setup();
+		executeDataSet(IMMUNIZATION_INITIAL_DATA_XML);
+	}
+		
+	@Test
+	public void shouldReturnExistingIMMUNIZATIONAsJson() throws Exception {
```
Comment (ibacher) [inline integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r4/ImmunizationFhirResourceProviderIntegrationTest.java:71]:
```suggestion
	public void shouldReturnExistingImmunizationAsJson() throws Exception {
```

**Review by ibacher — APPROVED** [review]

**Comment by ibacher** [comment]
@suruchee Thanks so much for this! Looks great!

**Comment by jonathandick** [comment]
Is this still active?

**Comment by ibacher** [comment]
@jonathandick This is just for an integration test. The resource itself already exists.

File: integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r4/ImmunizationFhirResourceProviderIntegrationTest.java [hunk 13]
```diff
@@ -0,0 +1,503 @@
...
+		
+		Immunization Immunization = readResponse(response);
+		
+		assertThat(Immunization, notNullValue());
+		assertThat(Immunization.getIdElement().getIdPart(), equalTo(IMMUNIZATION_UUID));
+		assertThat(Immunization, validResource());
+	}
+	
+	@Test
+	public void <span class="x x-first x-last">shouldReturnNotFoundWhenImmunizationNotFoundAsXML</span>() throws Exception {
```
Comment (Medhavi-16) [inline integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r4/ImmunizationFhirResourceProviderIntegrationTest.java:116]: @suruchee I think this line is causing a build fail.

File: integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r4/ImmunizationFhirResourceProviderIntegrationTest.java [hunk 14]
```diff
@@ -0,0 +1,503 @@
...
+
+import lombok.AccessLevel;
+import lombok.Getter;
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.lang3.time.DateUtils;
+import org.hl7.fhir.r4.model.Immunization;
+import org.hl7.fhir.r4.model.Bundle;
+import org.hl7.fhir.r4.model.CodeableConcept;
+import org.hl7.fhir.r4.model.DiagnosticReport;
+import org.hl7.fhir.r4.model.Enumeration;
```
Comment (Medhavi-16) [inline integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r4/ImmunizationFhirResourceProviderIntegrationTest.java:37]: These imports are unused. Please remove them.

File: integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r4/ImmunizationFhirResourceProviderIntegrationTest.java [hunk 15]
```diff
@@ -0,0 +1,503 @@
...
+import org.hl7.fhir.r4.model.OperationOutcome;
+import org.hl7.fhir.r4.model.Person;
+import org.junit.Before;
+import org.junit.Test;
+import org.openmrs.module.fhir2.BaseFhirIntegrationTest.FhirMediaTypes;
```
Comment (Medhavi-16) [inline integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r4/ImmunizationFhirResourceProviderIntegrationTest.java:42]: Unused import. You can run `mvn clean package` to remove all the unused imports.

**Comment by ibacher** [comment]
Closing this in favour of #343.

**Pull request closed without merging**